### PR TITLE
fix: unboundlocalerror when trying to retrieve attention feature maps using Qwen2FlashAttention2().forward()

### PR DESCRIPTION
**Pull request by EigenTom**
# What does this PR do?

Fix "unboundlocalerror" occurrs when trying to retrieve attention layer feature maps for models using `Qwen2FlashAttention2().forward()` to inference such as LLaVA-NeXT:

image

The main issue is in this method, the logic for computing `attn_weights` when flag `output_attentions` is set to `True` was not implemented, therefore leaving variable `atten_weights` unbounded.

Fixes # (issue)

Following code demonstrated in Qwen2Attention(nn.Module).forward(), several lines were added to compute the value of variable `atten_weights` when `output_attentions` is set to `True`:

image

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by EigenTom** [comment]
Replication Code:

~~~python3
import sys
sys.path.append(<path_to_LLaVA-NeXT>)
import os
os.environ["CUDA_VISIBLE_DEVICES"] = "0"

from llava.model.builder import load_pretrained_model
from llava.mm_utils import get_model_name_from_path, process_images, tokenizer_image_token
from llava.constants import IMAGE_TOKEN_INDEX, DEFAULT_IMAGE_TOKEN, DEFAULT_IM_START_TOKEN, DEFAULT_IM_END_TOKEN, IGNORE_INDEX
from llava.conversation import conv_templates, SeparatorStyle
from PIL import Image
import requests
import copy
import torch

import sys
import warnings

class LLAVANeXTService:
    def __init__(self, pretrained_model="lmms-lab/LLaVA-Video-7B-Qwen2", device="cuda:0", device_map="auto"):
        # Define the model loading arguments
        self.llava_model_args = {
            "multimodal": True,
        }
        overwrite_config = {}
        overwrite_config["image_aspect_ratio"] = "pad"
        self.llava_model_args["overwrite_config"] = overwrite_config

        # Load the tokenizer, model, and image processor
        self.tokenizer, self.model, self.image_processor, self.max_length = load_pretrained_model(
            pretrained_model, 
            None, 
            "llava_qwen", 
            device_map=device_map, 
            **self.llava_model_args,
            # attn_implementation="sdpa"
        )

        # Set model to evaluation mode
        self.model.eval()

        # Define the device (CUDA or CPU)
        self.device = torch.device(device)

    def predict(self, prompt, img_list,):
        # Load and process multiple images
        images = [Image.open(img_path) for img_path in img_list]
        image_tensors = process_images(images, self.image_processor, self.model.config)
        image_tensors = [_image.to(dtype=torch.float16, device=self.device) for _image in image_tensors]
        image_sizes = [image.size for image in images]

        # Prepare interleaved text-image input
        conv_template = "qwen_1_5"
        conv = copy.deepcopy(conv_templates[conv_template])
        
        if len(img_list) > 1:
            prompt = f"Provided are two images: {DEFAULT_IMAGE_TOKEN} and {DEFAULT_IMAGE_TOKEN} \n" + prompt
        else:
            prompt = f"Provided image is: {DEFAULT_IMAGE_TOKEN} \n" + prompt
        
        conv.append_message(conv.roles[0], prompt)
        conv.append_message(conv.roles[1], None)
        prompt_question = conv.get_prompt()

        # Tokenize the text prompt with image tokens
        inputs = tokenizer_image_token(prompt_question, self.tokenizer, IMAGE_TOKEN_INDEX, return_tensors="pt")
        input_ids = inputs.unsqueeze(0).to(self.device)

        # Generate the response from the model
        cont = self.model.generate(
            input_ids,
            images=image_tensors,
            image_sizes=image_sizes,
            do_sample=False,
            temperature=0,
            max_new_tokens=4096,
            return_dict_in_generate=True,
            output_attentions=True,
        )

        # Decode the output
        text_outputs = self.tokenizer.batch_decode(cont["sequences"], skip_special_tokens=True)
        # text_outputs = self.tokenizer.batch_decode(cont, skip_special_tokens=True)
        return text_outputs[0], cont

service = LLAVANeXTService()
text_output, raw_output = service.predict("what is the most salient object in the image", [<path_to_img>])
~~~
